### ISSUE TITLE
Add `exclude_if` (func) parameter to `model_dump` method

### ISSUE THREAD

**Issue by d3cryptofc** [pydantic/pydantic#10728]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

The `exclude_none` parameter is great!

However, I would like to be able to avoid even empty strings, empty dictionaries, empty lists, etc. And to avoid having to create a parameter for each type, I suggest adding the `exclude_by_func` parameter, where we can pass a function that will receive the value and validate whether it should remain in the dictionary or not.

Example:

**Args:**

- `instance`: Model instance, gives the user the power to know which model the field belongs to, in addition to being able to access attributes.
- `key`: Gives the user the field name.
- `value`: The value stored in the field.

```python
from pydantic import BaseModel

from types import NoneType
from typing import Any, Optional, Dict, List

def ignore_empty_values(instance: BaseModel, key: str, value: Any) -> bool:
   return isinstance(value, (list, dict, str, NoneType)) and not value

class Anything(BaseModel):
   a: Optional[int]
   b: Optional[List]
   c: Optional[Dict]
   d: Optional[str]
```

```python
>>> x = Anything(a=0, b=[], c={}, d='')
>>> # Dump all.
>>> x.model_dump()
{'a': 0, 'b': [], 'c': {}, 'd': '')
>>> # Dump but excluding empty values.
>>> x.model_dump(exclude_if=ignore_empty_values)
{'a': 0}
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [X] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by davidhewitt** [comment]
I literally just proposed a similar thing the other day in https://github.com/pydantic/pydantic-core/pull/1490#pullrequestreview-2384475109

(I called it `exclude_if`.)

**Comment by d3cryptofc** [comment]
Come to think of it, `exclude_if` is actually a better name for the parameter.

If this idea goes ahead, I will be saving an absurd amount of space in the database, avoiding all traces of empty data :)

**Comment by sydney-runkle** [comment]
This seems like a great idea to me, will help resolve many of the customization issues re exclude behavior 👍

PRs welcome with support!!

**Comment by d3cryptofc** [comment]
I made some changes to the issue description example, aiming to solve most of the problems.

> **Args:**
> - `instance`: Model instance, gives the user the power to know which model the field belongs to, in addition to being able to access attributes.
> - `field`: Gives the user all the information about the field.
> - `value`: The value stored in the field.
> ```python
> def ignore_empty_values(instance: BaseModel, field: FieldInfo, value: Any) -> bool:
>   return isinstance(value, (list, dict, str, NoneType)) and not value
> ```
Is it the right way?

**Comment by d3cryptofc** [comment]
Or an alternative using `&` and `|`:

```python
excludes: Callable = (
    (
        # Just if model is Phone or Address.
        JustForModel(Phone, Address),
        # And
        & (
            # Exclude fields with None value.
            exclude_none
            # Or then, exclude unsetted fields.
            | exclude_unset
            # Or then, exclude fields are using the default value.
            | exclude_defaults
        )
    )
    # Or
    | (
       # Just if model is Person
       JustForModel(Person)
       # And then, exclude fields with None value.
       & exclude_none
   )
)

person.model_dump(exclude_if=excludes)
```

```python
# Creating an exclude validator inheriting directly from the base class.
class JustForModel(ExcludeValidator):
    def __init__(self, models: BaseModel | Iterable[BaseModel]):
        self._models = models
    def __call__(self, instance: BaseModel, _, __):
        return isinstance(instance, self._models)

# Creating a function that transforms a function into a class that already inherits from the base class.
def fn_to_exclude_validator(function: Callable):
    return type(
        'ExcludeValidator',
        (ExcludeValidator,),
        dict(__call__=staticmethod(function))
    )()

# Usage example:
exclude_none = fn_to_exclude_validator(lambda _, __, value: value is None)
exclude_defaults = fn_to_exclude_validator(lambda instance, key, value:  instance.model_fields[key].default == value)
exclude_unset = fn_to_exclude_validator(lambda instance, key, _: key not in instance.__fields_set__)
```

**OR and AND class operators:**
```python
class ExcludeValidatorOr:
    def __init__(self, nodes: List[Callable]):
        self._nodes = nodes
    
    def __call__(self, instance: BaseModel, key: str, value: Any):
        for function in self._nodes:
            if function(instance, key, value):
                return True
              
    def __or__(self, other):
        nodes = self._nodes.copy()
        nodes.append(other)
        return type(self)(nodes)
    
    def __and__(self, other): return ExcludeValidatorAnd([self, other])

class ExcludeValidatorAnd:
    def __init__(self, nodes: List[Callable]):
        self._nodes = nodes

    def __call__(self, instance: BaseModel, key: str, value: Any):
        for function in self._nodes:
            if not function(instance, key, value):
                return False
            return True
    
    def __and__(self, other):
        nodes = self._nodes.copy()
        nodes.append(other)
        return type(self)(nodes)

    def __or__(self, other): return ExcludeValidatorOr([self, other])
```

**Validator base class:**
```python
class ExcludeValidator(ABC):
    @abstractmethod
    def __call__(self, instance: BaseModel, key: str, value: Any): ...
    def __or__(self, other): return ExcludeValidatorOr([self, other])
    def __and__(self, other): return ExcludeValidatorAnd([self, other])
```

**Comment by andresliszt** [comment]
I think it would be easier to implement at field level in the Field class, so the callable acts on a single field. I see a global callable acting on all the fields at once returning a boolean, a bit complicated, because all fields should evaluate a similar property, which is difficult when there are a large variety of types

**Comment by andresliszt** [comment]
Nice! Happy to help with if you are not planning to take it

**Comment by d3cryptofc** [comment]
although this will take away the flexibility in being able to choose different filters when necessary, for example, perhaps there is also the scenario in which the dump must be rerun completely to avoid problems with exclude_unset in the future (which fortunately is not my case, but it can to happen)

**Comment by davidhewitt** [comment]
We added `exclude_if` in https://github.com/pydantic/pydantic-core/pull/1535 - just need to document here.

**Comment by davidhewitt** [comment]
Renamed the title to `exclude_if`.

**Comment by davidhewitt** [comment]
Was documented in #12141

**Comment by Viicos** [comment]
@davidhewitt The issue mentions using `exclude_if` on the serialization methods (such as `model_dump()`). While I think the existing `exclude_if` at the field level makes more sense, I feel like we should keep this one open to track upvotes?

**Comment by davidhewitt** [comment]
🤦 Thank you

**Comment by davidhewitt** [comment]
That said, I'm probably not in favour of having this at the `model_dump()` level. That will be quite a performance hit, and _probably_ in most cases the field level will be good enough (?)

**Comment by Viicos** [comment]
Agree, we can leave it open for a bit for user input (especially after they can experiment with `exclude_if` at the field level in 2.12).

**Comment by d3cryptofc** [comment]
I also agree, but that means I would always be stuck with the same excludes, so if I want a dump with other excludes it won't be possible.

Like for example a dump without exclude, maybe `model_dump(ignore_excludes=True)` would be good, It doesn't resolve other multiple excludes scenarios but resolves one of them.

**Comment by sam-mosleh** [comment]
Thanks for the great work on v2.12! 🎉

@davidhewitt I have a question about the current design of `exclude_if` – hopefully this is the right place to ask. Based on the current implementation, `exclude_if` only evaluates based on field values, which means we can't replicate the behavior of `exclude_unset` (I'm thinking about a single `mode_dump()` interface by ditching different exclude options entirely).
Would it be possible to support a more advanced Rust-based alternative [similar to the approach](https://github.com/pydantic/pydantic/issues/10728#issuecomment-2451018023) proposed by @d3cryptofc? This would enable more sophisticated exclusion logic beyond just value-based filtering.

Something like this:
```python
class SomeModel(pydantic.BaseModel):
    a: Annotated[int, pydantic.Exclude(pydantic_core.exclude_unset), pydantic.Exclude(pydantic_core.exclude_if_equal(0))] = 1
```

**Comment by Viicos** [comment]
> which means we can't replicate the behavior of `exclude_unset` (I'm thinking about a single `mode_dump()` interface by ditching different exclude options entirely).
> Would it be possible to support a more advanced Rust-based alternative [similar to the approach](https://github.com/pydantic/pydantic/issues/10728#issuecomment-2451018023) proposed by [@d3cryptofc](https://github.com/d3cryptofc)? This would enable more sophisticated exclusion logic beyond just value-based filtering.

I think what I proposed in https://github.com/pydantic/pydantic/issues/12056 is quite similar: markers than can be combined together and would be understood natively by pydantic-core (you'll see examples using `exclude_if`). It could apply to a model-level `exclude_if` option.
